### ISSUE TITLE
pluginHost stackType may be useless

### ISSUE THREAD

**Issue by hugoaboud** [monocasual/giada#174]
Hi, I've recently found about giada and forked it to implement a few features (here's the roadmap of my fork: https://github.com/hugoaboud/giada/issues/1).

I basically want to implement multichannel input, and as I was digging into the pluginList code I found that pluginHost uses getStack to choose between two master stacks or the channel one, and I think that it would be way cleaner to have a "MasterChannel" extending channel and remove all references to stackType from code.
What do you think about? I'm implementing both InputChannel and MasterChannel on my fork, when done I'll push it and post here.

**Comment by monocasual** [comment]
Hi @hugoaboud , pluginHost:: is quite "primitive" if compared to other code areas. I agree the getStack function could be refactored into something smarter than its current form. 

Just a little word of caution: the entire "core" code base is going through [a deep refactoring](https://github.com/monocasual/giada/tree/opaque-channels) - we are switching to a smarter audio buffer instead of raw pointers for handling input and output data, and pluginHost will be affected by such changes. Maybe you want to wait a couple of weeks before digging into it...

That said, +1 for the roadmap. Let me know if you need more insights to get stated.

**Comment by hugoaboud** [comment]
Well, this refactor kinda gets me stuck for now, but I can't take weeks apart from my project, so I guess I'll keep implementing my features and then spend a little more time later merging the refactored core.
I've pushed a commit with a simple "Edit > Input channels" dialog If you'd like to take a look, and I'm starting to change the Mixer core so it has 3 channel stacks instead of one: input, columns and master. Column channels should have their own samplechannel stack.

What really got me interested on giada is the modularity of your concept of channels, and the hability to create/destroy them live. I come from an Ableton Live environment and when I've tried to do something alike in it with python Midi Scripts things got terribly buggy.

Because of that modularity I'm not really sure If Sample/midi Channels should necessairly exist within a column channel (ie the Mixer stack should store columns, samples or both). I need your opinion about It.
Graphically there's no way a sample exists outside a column, so it feels resonable that the mixer stack has column channels and each column has a sample channel stack.
That way all sample inside a column should receive audio from the input channels routed to that column and output to the column buffer.

Basically what I conceive is: I've got a guitar plugged in line and a mic for trumpet and percussion sounds. So I create 4 input channels for example: guitar, trumpet, perc1 and perc2. Both the perc and the trumpet source their audio from the same Jack channel. Each input has a separate VST chain for specific tones.
Then I have 3 columns: guitar, trumpet and percussion. Both the perc input channels are routed to the same column.
Each column has a VST chain that allows me to apply filtering and modulation to the playing samples of that column.
That way I can easily create and manage layers live with a musical sense of grouping (like having two perc inputs to the same column) and arranging gets easier.

**Comment by hugoaboud** [comment]
Just a small update: I've started the hard work on the multichannel feature. So far I've comitted/pushed a "Edit > Input Channels" dialog that allows the user to manipulate the mixer's vector of InputChannels.
I've also modified mixer masterPlay so that input peak and monitoring are handled by InputChannel process (the inBuf is stored deinterleaved by InputChannels that are either being used for recording or have their inputMonitor activated). Selecting an input (still fixed to 2) on the dialog and activating/deactivating the input monitor are working already.
Now moving on in the signal chain...

**Comment by hugoaboud** [comment]
Also, could you explain me the point of this loop on _mixer::masterPlay_?

- Why do you call processLineIn, lineInRec and sumChannels frame by frame and not on the entire buffer at once? (I understand you update the clock current frame, but I don't see it being used in any of those methods)
- I've changed the RtAudio input parameters so there can be more than 2 inputs, how does it affect doQuantize, testBar, etc. if j=0,3,6,9,... instead of j=0,2,4,6... ?

```cpp
for (unsigned j=0; j<bufferSize; j+=2) {
	processLineIn(inBuf, j);
	if (clock::isRunning()) {
		lineInRec(inBuf, j);
		doQuantize(j);
		testBar(j);
		testFirstBeat(j);
		readActions(j);
		clock::incrCurrentFrame();
		testLastBeat();  // this test must be the last one
		clock::sendMIDIsync();
	}
	sumChannels(j);
}
renderIO(outBuf, inBuf);
```

**Comment by hugoaboud** [comment]
Another question: wouldn't it be cleaner to always use conf::buffersize instead of storing it on every channel? Is there a reason for that?

**Comment by monocasual** [comment]
Looks good ;) Regarding your questions:

1. "Why do you call processLineIn, lineInRec and sumChannels frame by frame" - It's kind of meaningless, I agree. There's room for optimization in there, like moving `lineInRec` and other functions outside the for loop. Other functions need to be processed frame by frame, e.g. `readActions()`, since every frame might hold useful data;
* "I've changed the RtAudio input parameters so there can be more than 2 inputs" - that's the mess we are going to fix right now in the development branch. With the "vanilla" version you are working with you have to take care of channels as well. So j=0,2,4,6,... would become j=0,4,8,12, ...case of a 4-channels input scenario.

**Comment by hugoaboud** [comment]
Thanks! I've spent a few more days coding and I'm almost done implementing the per channel recording (instead of recording to vChanInput and then copying to the channel wave), should push it until the weekend.
Then I'll close this issue and open a issue named "Multichannel support" so people can test my fork and help me find bugs. When opaque channels are done I'll merge them into my fork. Sounds good?

I've changed a little the _InputRec Button_ behavior:
   - **why:**  I like the hability to arm a few SampleChannels and then press record to record them all simultaneously (this goes pretty well with multichannel support). But I'd also like to be able to manipulate recordings individually within a performance.
   - **how:**  I've removed the alert for "no channels armed" on the InputRec Button. If there are armed channels it triggers record on those channels, else it just sets _mixer::recording_ to true. When arming a channel, if _mixer::recording_ is false it gets armed as usual, if it's true the channels is triggered for recording. When InputRec Button goes off it triggers rec stop on every recording channel.
   - **compatibility:** The current usability is kept (starting and stopping multiple channel recordings simultaneously with the InputRec button) but if the user keeps InputRec on, the mode changes. It can also be good for musical performance to have both methods integrated that way.
   - **additional:** Now that recording is handled by channel I was thinking it might be interesting to have it behaving differently for each loop mode. For example: _loop.basic_ waits for the sequencer and overdubs infinitely; _loop.single_ waits for the sequencer but records a single layer; _oneshot.basic_ starts recording immediately and stops when called again; _oneshot.press_ records immediately and stops when user releases button.

What do you think?

**Comment by hugoaboud** [comment]
Also, on this recording refactoring I've realized it'd be good to encapsulate _Channel_'s methods and parameters regarding preview, play, record, and anything related to content (actions included) to a _ContentChannel_ class.
That differs them from _Input_, _Column_ and _Master_ raw _Channels_ that don't store content, only buffers.
That way I can safely create and use a polymorphic ContentChannel stack on ColumnChannel. Does it sound reasonable?

**Comment by monocasual** [comment]
Plug-in processing architecture changed, improved and simplified in 0.16.0. Closing this.
